### PR TITLE
Stats: Removing hard-coded tiers

### PR DESCRIPTION
**Pull request by grzegorz-cp**
Related to #84534

## Proposed Changes

* updating tier slider with values and prices from the new endpoint
* adding a popover explaining the highest tier pricing and extension

## Testing Instructions

* navigate to a live branch
* apply the feature flag `stats/tier-upgrade-slider`
* in case you are viewing stats for a non-commercial website click on the link allowing the purchase of a commercial licence
* check various pricing options and verify that they display prices in your currency
* check that the highest tier also shows a popover with additional information

After the changes:
SCR-20231129-nzix

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

### CODE REVIEW

File: client/my-sites/stats/stats-purchase/stats-purchase-tier-upgrade-slider.tsx
Comment (grzegorz-cp) [inline client/my-sites/stats/stats-purchase/stats-purchase-tier-upgrade-slider.tsx:82]: I didn't want to hard-code this, but adding a loop that finds the highest tier and multiplies the limit felt over the top as well. What do you think about this solution?

**Comment by a8ck3n** [comment]
These changes look pretty good to me. I'd prefer if we held off on merging this until the API bits are merged though, otherwise it forces us to sandbox to continue working on the visual components. Just running from the Live branch, I currently see this:

SCR-20231129-obnj

This also makes me think we need to better handle a potential empty state in the slider/wrapper so users don't see this if they have a network call time-out or fail.

File: client/my-sites/stats/stats-purchase/stats-purchase-tier-upgrade-slider.tsx
Comment (a8ck3n) [inline client/my-sites/stats/stats-purchase/stats-purchase-tier-upgrade-slider.tsx:84]: Could we move this logic up to `stats-purchase/index.tsx` or `stats-purchase/stats-purchase-single-item.tsx`?

My thinking was we'd do the page-specific translation logic at a higher level and just pass in the already translated/messaged strings and values to the component. In that way, we can easily use this same component for the PWYW flow.

If we put this logic directly in `TierUpgradeSlider` we'd need to duplicate the component for the updated PWYW flow. If we abstract the logic a bit, we could call the slider with some already translated UI strings, some "steps", and a callback and use it for both flows.

So instead of:

```
<TierUpgradeSlider priceTiers={ priceTiers } currencyCode={ currencyCode } />
```

…something like:

```
<TierUpgradeSlider
  lhCallout="Monthly site views limit"
  rhCallout="You will pay"
  infoMessage="Price per month, billed yearly"
  onChange={ handleOnChange }
  steps={ paidUpgradeTiers }
/>
```

Then for PWYW we can do almost the same thing:

```
<TierUpgradeSlider
  lhCallout="Your monthly contribution"
  rhCallout="Thank you!"
  infoMessage="The average person pays $5 per month, billed yearly"
  onChange={ handleOnChange }
  steps={ pwywUpgradeTiers }
/>
```

We'd then need to define shape of the `steps` data but maybe just an array of `{ lhValue, rhValue, infoMessage }` values or something similar?

**Review by a8ck3n — COMMENTED** [review]
Put most of my thoughts into the inline comments. Let me know what you think of abstracting the logic up a level or two though.

My main concern at the moment is that we not merge this until the API support lands (even if just a limited version) so as to keep the test environment simple.

**Comment by kangzj** [comment]
>  I'd prefer if we held off on merging this until the API bits are merged though, otherwise it forces us to sandbox to continue working on the visual components. @a8ck3n 

That's a valid concern; however, the API bits are merged already, but only the new pricing isn't live. Switching the new pricing to live would be the last step for us when everything else is ready. If you find the sandbox method is not very handy, try to use the cookie method, where you wouldn't need your sandbox anymore. In short, I would appreciate PRs continue to merge if that makes sense 😄 

-----Update-------

Keep the mock data might be a good idea as a compromise tho! 🤔

File: client/my-sites/stats/stats-purchase/stats-purchase-tier-upgrade-slider.tsx
Comment (kangzj) [reply client/my-sites/stats/stats-purchase/stats-purchase-tier-upgrade-slider.tsx:84]: If it's just translations then it makes sense to do so, but when I think about it, there are marks, emojis and probably more differences for the two views. I guess we could do two atm, and refactor when we see a lot of duplication.

What do you say? @grzegorz-cp  @a8ck3n

**Review by kangzj — APPROVED** [review]
It works well for me. Good job! 👍 

IMHO @a8ck3n's point sounds reasonable - should we keep the mock data as fallback instead of removing it now?

File: client/my-sites/stats/stats-purchase/stats-purchase-tier-upgrade-slider.tsx
Comment (grzegorz-cp) [reply client/my-sites/stats/stats-purchase/stats-purchase-tier-upgrade-slider.tsx:84]: That's a great argument, and it made me go through the structure one more time.
I'm open to modification ideas, but after looking at the designs again, I think the PWYW slider and Commercial slider only share CSS, but they are actually not that similar.

TieredUpgradeSlider is already an abstraction layer over the Slider component, and I believe that the tier manipulation belongs there because it takes incoming data and prepares it for the format needed by the slider. Other components don't need that structure. 

I'm afraid that if we start moving that logic to the higher components, those components will start to grow and be hard to maintain - there will be more logic coming from other children to match the pattern. On the other hand, if we move the logic up and also update the TieredUpgradeSlider to handle multiple cases, we end up inflating that component as well (e.g. marks vs emoji, showing/hiding slider for the extension tiers, etc.). In the end, we'll get something similar to the StatsModule components that are on the grid below the chart of the traffic page. One component handles all of the different widgets but each widget is slightly different.

File: client/my-sites/stats/stats-purchase/stats-purchase-tier-upgrade-slider.tsx
Comment (grzegorz-cp) [inline client/my-sites/stats/stats-purchase/stats-purchase-tier-upgrade-slider.tsx:102]: I restored mock data, but it will be available locally - not through the live branch. This way even if we forget to remove the mock, the component will work as it needs to and we can test various currencies on the the live branch.

**Review by kangzj — APPROVED** [review]
:shipit:

**Comment by kangzj** [comment]
One nitpick I notice while testing is that when the slider the the right most, and the popover shows up which would cause the slider to lose focus. This is a low priority issue and I'm happy for us to fix later.

**Comment by a8ci18n** [comment]
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/9802586

Some locales (Japanese) have been temporarily machine-translated due to translator availability. All other translations are usually ready within a few days. Untranslated and machine-translated strings will be sent for translation next Monday and are expected to be completed by the following Friday.

Thank you @grzegorz-cp for including a screenshot in the description! This is really helpful for our translators.

**Comment by a8ci18n** [comment]
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/9802586).

**Pull request merged**
